pandas/_libs/groupby_helper.pxi.in

```diff
@@ -588,7 +588,13 @@ def group_rank_{{name}}(ndarray[float64_t, ndim=2] out,
                 if out[i, 0] != out[i, 0] or out[i, 0] == NAN:
                     out[i, 0] = NAN
                 else:
-                    out[i, 0] = out[i, 0] / grp_sizes[i, 0]
```

Comment:
you need to fill with NAN if the grp_sizes[i, 0] == 0

Reply:
make this another elif condition, then the else can be the division.

Reply:
Aren't lines 588-589 taking care of the NAN edge case?

In my testing I have only ever been able to get grp_size[i, 0] == 0 in the bugged case (pct=True, method="dense").

So my idea of the fix was that grp_size[i, 0] should never be 0 in the else block and if it is, then it should be 1 (not modifying out[i, 0] would be the same as div by 1).

Reply:
no, you need to add this case there as well, if the denominator is 0, the result is NaN

Reply:
As it is right now, doing that would cause incorrect results (would fail the test I added in test_rank.py). The reasoning behind this is:

1. grp_size[i, 0] should legitimately be 0 only if all the values in a group are NaN. In this case the if statement in Lines 588-589 takes care of this case.

2. grp_size[i, 0] shouldn't be 0 but is in the case where there is a group with only one member with a non NaN value but another index also has the same value but in a different group && method="density". In this case the correct grp_size[i, 0] should be 1 and it being 0 is what causes BUG 23666.

Therefore my fix is only performing the division by grp_size[i, 0] when it is not equal to 0 in the if block [Line 591] because the else condition only demands division by 1 (the corrected grp_size[i, 0]).

To summarise, making the result NaN on grp_size[i, 0] would stop the ZeroDivisionError from occurring in BUG 23666 and replace it with incorrect behaviour.

An additional note: the conclusion of my testing is that there is a flaw in the calculation of grp_sizes[i, 0] itself under these circumstances, but in my testing I have uncovered that there are other parameters (not just method="dense") that can have weird behaviour that is caused due to incorrect grp_sizes[i, 0]. Once, I'm able to determine the extent of this I'm going to create a new issue that would target the fixing of grp_size[i, 0] calculation bug. This fix is a stop-gap measure till then since this would correct one instance of incorrect rank behaviour.

Reply:
Please improve the test case to make an assertion about the result of this expression. It’s not enough to assert that it simply computes which is the behavior now. 

Doing that would aid this discussion

Reply:
Ah sorry, my bad. I forgot to push it.